src/Microsoft.ML.DataView/DataViewType.cs

```diff
@@ -461,4 +480,48 @@ public override bool Equals(DataViewType other)
 
         public override string ToString() => "TimeSpan";
     }
+
+    /// <summary>
+    /// <see cref="DataViewTypeAttribute"/> should be used to decorated class properties and fields, if that class' instances will be loaded as ML.NET <see cref="IDataView"/>.
+    /// The function <see cref="Register"/> will be called to register a <see cref="DataViewType"/> for a <see cref="Type"/> with its <see cref="Attribute"/>s.
+    /// Whenever a value typed to the registered <see cref="Type"/> and its <see cref="Attribute"/>s, its <see cref="DataViewSchema.Column.Type"/> would be the
+    /// associated <see cref="DataViewType"/>.
+    /// </summary>
+    public abstract class DataViewTypeAttribute : Attribute
+    {
+        /// <summary>
+        /// A function implicitly invoked by ML.NET when processing a custom type. It binds a DataViewType to a custom type plus its attributes.
+        /// </summary>
+        public abstract void Register();
+
+        /// <summary>
+        /// Return <see langword="true"/> if <see langword="this"/> is equivalent to <paramref name="other"/> and <see langword="false"/> otherwise.
+        /// </summary>
+        /// <param name="other">Another <see cref="DataViewType"/> to be compared with <see langword="this"/>.</param>
+        public abstract bool Equals(DataViewTypeAttribute other);
```

Comment:
This pattern isn't a typical .NET pattern. I think the derived classes should just override `Equals` and `GetHashCode`, just like any typical .NET class.